htmlproofer/plugin.py

```diff
@@ -59,8 +68,54 @@ def get_url_status(self, url: str, soup: BeautifulSoup) -> Tuple[str, int]:
             except requests.exceptions.ConnectionError:
                 return clean_url, -1
         else:
+            match = re.match(r'(.+)#(.+)', clean_url)
+            if match is not None:
+                # URL is a link to another local Markdown file that includes an anchor.
+                url_target, anchor = match.groups()
+                target_markdown = self.find_target_markdown(url_target, files)
+                if (target_markdown is None
+                        or not self.contains_anchor(target_markdown, anchor)):
+                    # The corresponding Markdown header for this anchor was not found.
+                    return url, 404
+
             return url, 0
 
+    @staticmethod
+    def find_target_markdown(url: str, files: Files) -> Optional[str]:
+        """From a built URL, find the original Markdown source from the project that built it."""
+        # Remove first / from absolute URLs to match how MkDocs stores URLs in its Files.
+        url = url.lstrip("/")
+
+        for file in files.src_paths.values():  # type: File
+            # Using endswith() is to deal with relative URLs that do not contain the full path
+            # to the .html file. This approximation will allow a small number of anchors to be
+            # validated even if they don't exist (if the same Markdown filename is used in
+            # multiple folders), but the alternative is to try to reimplement MkDocs file/URL
+            # generation.
+            if file.url.endswith(url):
+                return file.page.markdown
+        print(f"Warning: Unable to locate Markdown source file for: {url}", file=sys.stderr)
+        return None
+
+    @staticmethod
+    def contains_anchor(markdown: str, anchor: str) -> bool:
+        """Check if a set of Markdown source text contains a heading that corresponds to a
+        given anchor."""
+        for line in markdown.splitlines():
+            # Markdown allows whitespace before headers and an arbitrary number of #'s.
+            match = re.match(rf'\s*#+\s*(.*)', line)
+            if match is not None:
+                heading = match.groups()[0]
+
+                # Headings are allowed to have images after them, of the form:
+                # # Heading [![Image][image-link]]
+                # But these images are not included in the generated anchor, so remove them.
+                heading = heading.split("[")[0].strip()
```

Comment:
do we have a test for this case ?

Reply:
Yes, the link to the top level header in index.md is an example of this https://github.com/johnthagen/mkdocs-htmlproofer-plugin/blob/cross-page-anchor-links/test/docs/index.md#mkdocs-htmlproofer-plugin-

https://raw.githubusercontent.com/manuzhang/mkdocs-htmlproofer-plugin/c9545e7d93ef9e779f059ab27bb373954aed7a81/test/docs/nested/page1.md